src/System.Text.Json/tests/JsonReaderTests.cs

```diff
@@ -8,153 +7,252 @@ namespace System.Text.Json.Tests
                     case JsonReader.JsonTokenType.ArrayEnd:
-                        json.ReadArrayEnd();
-                        property = new Utf8String("");
-                        value = null;
                         break;
-                    case JsonReader.JsonTokenType.PropertyName:
-                        property = json.ReadPropertyAsString();
-                        value = null;
+                    case JsonReader.JsonTokenType.Pair:
+                        var pair = new Pair
+                        {
+                            Name = (string) jsonReader.ReadString(),
+                            Value = GetValue(ref jsonReader)
+                        };
+                        if (jsonPairs != null) jsonPairs.Add(pair);
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsString:
-                        value = json.ReadPropertyAsString();
+                    case JsonReader.JsonTokenType.Value:
                         break;
-                    case JsonReader.JsonTokenType.PropertyValueAsInt:
-                        value = json.ReadPropertyValueAsInt();
+                    case JsonReader.JsonTokenType.Finish:
                         break;
                     default:
-                        property = new Utf8String("");
-                        value = null;
-                        break;
+                        throw new ArgumentOutOfRangeException();
                 }
-                if (property != new Utf8String("") && value != null)
+            }
+            throw new FormatException("Json object was started but never ended.");
+        }
+
+        private static Array ReadArray(ref JsonReader jsonReader)
+        {
+            Array jsonArray = null;
+            List<Value> jsonValues = null;
+
+            while (jsonReader.Read())
+            {
+                switch (jsonReader.TokenType)
                 {
-                    jsonOutput.Add(new KeyValuePair<Utf8String, object>(property, value));
+                    case JsonReader.JsonTokenType.Start:
+                        break;
+                    case JsonReader.JsonTokenType.ObjectStart:
+                        break;
+                    case JsonReader.JsonTokenType.ObjectEnd:
+                        break;
+                    case JsonReader.JsonTokenType.ArrayStart:
+                        jsonArray = new Array();
+                        jsonValues = new List<Value>();
+                        break;
+                    case JsonReader.JsonTokenType.ArrayEnd:
+                        if (jsonArray != null)
+                        {
+                            var jsonElements = new Elements {Values = jsonValues};
+                            jsonArray.Elements = new List<Elements> {jsonElements};
+                            return jsonArray;
+                        }
+                        break;
+                    case JsonReader.JsonTokenType.Pair:
```

Comment:
A pair is a 'member' of an object and is of the form string:value.

http://json.org/
> An object is an unordered set of name/value pairs. An object begins with { (left brace) and ends with } (right brace). Each name is followed by : (colon) and the name/value pairs are separated by , (comma).